psyneulink/library/compositions/autodiffcomposition.py

```diff
@@ -528,16 +556,20 @@ def execute(self,
             # FIX 5/28/20:
             context.execution_phase = execution_phase
 
-            report(self,
-                   # [LEARN_REPORT],
-                   [EXECUTE_REPORT, PROGRESS_REPORT],
-                   report_num=report_num,
-                   scheduler=scheduler,
-                   content='trial_end',
-                   context=context)
-
-            scheduler.get_clock(context)._increment_time(TimeScale.TRIAL)
-
+            if report is not None:
+                report(self,
+                    # [LEARN_REPORT],
+                    [EXECUTE_REPORT, PROGRESS_REPORT],
+                    report_num=report_num,
+                    scheduler=scheduler,
+                    content='trial_end',
+                    context=context)
+
+            # FIXME: In nested autodiff, the scheduler is unable to correctly update the timescale
+            try:
+                scheduler.get_clock(context)._increment_time(TimeScale.TRIAL)
+            except KeyError as e:
+                pass
```

Comment:
```suggestion
            scheduler.get_clock(context)._increment_time(TimeScale.TRIAL)
```
I think this would throw KeyError when run outside of nesting, right? In which case I think we want to have a clock for context and run this line every time:

```diff
diff --git a/psyneulink/library/compositions/autodiffcomposition.py b/psyneulink/library/compositions/autodiffcomposition.py
index fc5c948c82..1cf0ada9a3 100644
--- a/psyneulink/library/compositions/autodiffcomposition.py
+++ b/psyneulink/library/compositions/autodiffcomposition.py
@@ -518,6 +518,11 @@ class AutodiffComposition(Composition):
         if scheduler is None:
             scheduler = self.scheduler
 
+        try:
+            scheduler.get_clock(context)
+        except KeyError:
+            scheduler._init_counts(context.execution_id, self.base_execution_id)
+
         if self._is_learning(context):
             # TBI: How are we supposed to use base_context and statefulness here?
             # TBI: can we call _build_pytorch_representation in _analyze_graph so that pytorch
```

Reply:
I don't think it needs to be avoided because autodiff doesn't call Scheduler.run like other compositions, but I think it's better to do it as in the diff and suggestion. Otherwise, in the case the `get_clock(context)` call fails, there just won't be a time for context, rather than a time accurately reflecting the number of trials that were run.